### ISSUE TITLE
Publish on packagist?

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#55]
With some more extensions to the composer.json we could publish PrivateBin on https://packagist.org/.

However I am not sure whether this is the best place as composer (and packagist) are more used for libraries and not for complete-web projects as PrivateBin AFAIK. (Correct me if I am wrong.)

**Comment by elrido** [comment]
With the addition of the composer.json it is a logical step to eventually publish it on packagist. We could also look into other web application package managers like [sandstorm](https://docs.sandstorm.io/en/latest/vagrant-spk/packaging-tutorial/). And of course all the classics: Debian, Ubuntu, RHEL/CentOS and company.

**Comment by rugk** [comment]
Yeah have not thought about sandstorm for example. It works be great to have it published there too. Don't know what we have to adjust, but I think at least for sandstorm you need some code changes as it e.g. wants to separate each individual Bins somehow.

**Comment by rugk** [comment]
Yeah have not thought about sandstorm for example. It works be great to have it published there too. Don't know what we have to adjust, but I think at least for sandstorm you need some code changes as it e.g. wants to separate each individual Bins somehow.

**Comment by rugk** [comment]
Yeah have not thought about sandstorm for example. It works be great to have it published there too. Don't know what we have to adjust, but I think at least for sandstorm you need some code changes as it e.g. wants to separate each individual Bins somehow.

**Comment by rugk** [comment]
Yeah have not thought about sandstorm for example. It works be great to have it published there too. Don't know what we have to adjust, but I think at least for sandstorm you need some code changes as it e.g. wants to separate each individual Bins somehow.

**Comment by k0nsl** [comment]
lol @rugk , that's a lot of double-posts :P

**Comment by rugk** [comment]
Oh, ...

(deleted)

**Comment by Sobak** [comment]
@rugk 

> However I am not sure whether this is the best place as composer (and packagist) are more used for libraries and not for complete-web projects as PrivateBin AFAIK. (Correct me if I am wrong.)

You are not wrong. If we publish it on Packagist, users will be able to do `composer create-project`, but I doubt it is widely used.

**Comment by rugk** [comment]
Ah, good I am wrong. :smile:

**Comment by rugk** [comment]
Ah, no I am "not" wrong, mhh. Should we still publish it on Packagist? Just for the sake of doing so? I mean it should not hurt and when users feel like user Composer would be good, then they can do it.
The publication of Packagist happens automatically once it is setup.

**Comment by elrido** [comment]
We need to maintain the composer file anyway for one of the code checkers. So why not go the additional step and publish it on composer? And if a few server admins have an easier time installing it thanks to that, its worth it.

**Comment by rugk** [comment]
Okay, so I published it on Packagist: https://packagist.org/packages/privatebin/privatebin

**Comment by rugk** [comment]
We could also publish it on [Pear](https://pear.php.net/account-request-newpackage.php), but there they have some Coding standards requirements.

**Comment by Sobak** [comment]
No, please. PEAR is not dead (yet), but it definitely should be. It requires not only specific coding standard but many meta-files (`package.xml`) etc. It's pain in the ass, trust me and gives us no benefit.

**Comment by elrido** [comment]
I just found [someones dockerfile](https://hub.docker.com/r/wonderfall/zerobin/) for PrivateBin. I'm unfamiliar with Alpine Linux, but it seems to be a very compact distro (busybox, small libc) making optimal for container images. I'll contact the user and see if we could provide a version of this but maintained by us in our organization account (I'd like to change the version used to the latest release instead of master).

**Comment by dvzrv** [comment]
I'm interested in packaging this for Arch Linux. We currently have [a package for privatebin in the Arch User Repository (AUR)](https://aur.archlinux.org/packages/privatebin/), but I could potentially also bring this to [[community]](https://www.archlinux.org/packages/?sort=&repo=Community&q=&maintainer=&flagged=).
The current PKGBUILD still needs quite some improvements (e.g. separate user, systemd tmpfiles/sysusers integration, be FHS compliant), but I'm interested in using the application myself, so I'll do this in the coming months and get back to you!

**Comment by elrido** [comment]
That is very cool, thank you for considering it!

In case this information can help others, here are my notes from back when I had started looking into packaging it for Debian & CentOS (never took the time to finish it, unfortunately):
- most packaged web applications get installed into /usr/share, i.e. /usr/share/privatebin
- the application should be split into a folder that can be served in the webserver and the internal folders (incl. data), see [INSTALL.md](https://github.com/PrivateBin/PrivateBin/blob/master/INSTALL.md#changing-the-path) and the [Dockerfile](https://github.com/PrivateBin/docker-nginx-fpm-alpine/blob/master/Dockerfile#L51)
- the application path needs to have the correct SELinux type `httpd_sys_content_t` in CentOS
- the data folder should be a symlink to /var/lib/privatebin/data, as it contains variable data
- the data folder needs to have the correct SELinux type `httpd_sys_**rw**_content_t` in CentOS
- Ideally, the package would include default configuration snippets that make the application accessible on a /privatebin URL path for apache and nginx
- the various `*.md` files and the `cfg/conf.sample.php` could be moved to /usr/share/docs/privatebin
- commit .spec (RPM) or .dsc (deb) files to the repo
